php/libraries/NDB_BVL_Instrument.class.inc

```diff
@@ -2190,6 +2190,82 @@ abstract class NDB_BVL_Instrument extends NDB_Page
         }
     }
 
+    /**
+     * Loads multiple instruments representing multiple commentIDs from the
+     * database in bulk. This function must be called on an instrument instantiated
+     * as \NDB_BVL_Instrument::factory('instrumentname') without a comment ID, and
+     * will an array of copies of the instrument type with the data for the given
+     * commentIDs loaded.
+     *
+     * @param $commentIDs string[] A list of commentIDs to load in bulk.
+     *
+     * @return \NDB_BVL_Instrument[]
+     */
+    function bulkLoadInstanceData(iterable $commentIDs): array
+    {
+        if ($this->commentID !== null && $this->commentID !== '') {
+            throw new \LogicException(
+                "Must bulk load from instrument loaded without commentID"
+            );
+        }
+        $db = \Database::singleton();
+
+        $prepBindings = [];
+        $prepValues   = [];
+
+        $i = 0;
+        foreach ($commentIDs as $commentID) {
+            $i++;
+
+            $prepBindings[]      = ":cid$i";
+            $prepValues["cid$i"] = $commentID;
+        }
+
+        if ($this->jsonData) {
+            // We filter out if Data is null for 2 reasons.
+            // 1. Otherwise getInstanceData will re-query the database since
+            //    $newinst->instanceData is null and commentID is set
+            // 2. The Data column being null in the flag table is equivalent
+            //    to there not being a row in the table in the non-JSON
+            //    case (where it will also not get returned because there's
+            //    no row in the table.)
+            $jsondata = $db->pselect(
+                "SELECT CommentID, Data FROM flag WHERE Data IS NOT NULL" .
+                " AND CommentID IN ("
+                . join(',', $prepBindings) . ')',
+                $prepValues,
+            );
```

Comment:
Not sure if this makes a difference here at all, but it might later down the line i think.

given this approach you are getting different results:
1. between JSON instruments and SQL table instruments
2. between `bulkLoadInstanceData()` and `loadInstanceData()` 

because you are not querying any JSON data instrument which has no values saved, your results will actually look like the instrument never was populated in the flag table in the first place while for SQL table instruments you would actually get an entry in the array for the commentID and an associated array with the CommentID from the instrument table inside... making the returned result difficult to manage for any code calling this function because the return seems inconsistent.

Even loadInstanceData effectively returns an empty array for instruments with no Data (which is not exactly the same as SQL table instruments but still better than null).

ie. if commentID `123456MTL000999999` has no instrument data:
If it's a `jsonData=false` you get
```
[ 
    Instrument (CommentID:`123456MTL000999999`, data: [CommentID=>`123456MTL000999999`])
]
``` 
If its a `jsonData=true` you get
```
[]
```

I think there's value in matching the returns regardless of the type of data saving we decide to go for... or at least returning
```
[ 
    Instrument (CommentID:`123456MTL000999999`, data: [])
]
```

Reply:
It's been a while since I wrote this code so I'll have to investigate. From what I remember (because of the comment..) the problem was that getInstanceData starts

```
   public function getInstanceData(): array
    {
        if ($this->instanceData !== null) {
            return $this->instanceData;
        }
```

So if this code sets instanceData to null (because the column is null in the database) it doesn't, in fact, bulk load the data, but still runs an SQL statement per row and defeats the purpose of using this function.. but if that was the only problem an empty array should suffice.

Reply:
either empty array or an array with only the commentID as an element would do the trick I think and mimick what happens with SQL instruments i believe

Reply:
I don't think that's true. With SQL table instruments the data is loaded with

```
SELECT * FROM $this->table WHERE CommentID
```

And there is no row in `$this->table` unless data entry has been done. (At least according to point 2 in the comment, if that's still accurate.)

Reply:
I am 99.9 % sure the row in the instrument gets created when the battery is populated. along with the DDE_ row as well regardless if the instrument is DDE or not.

you give too much credit to our current instrument infrastructure

Reply:
The flag table gets populated by NDB_BVL_Battery->addInstrument (along with the DDE row, regardless of if it's a DDE instrument...) but not the table. The instrument table either gets populated the first time the instrument is accessed in the frontend or the first time data is saved, I can't remember which. (The latter would make more sense but something tells me it's the former..)

Reply:
I think the instrument table is populated when we start the stage.

Reply:
I dug a little more and addInstrument does insert into the table unconditionally, I just missed it the first time I was looking at it. It appears to have never been updated to use json instrument data. So from what I can tell (Ihaven't tested yet..) you can't currently include a json data instrument in a battery.

Reply:
@laemtl is that true? have you tried start next stage on jsonData instruments ?

Reply:
Yes

Reply:
How does it get past this line in addInstrument?

```
    // instantiate instrument object to get table name attribute
        $obj = NDB_BVL_Instrument::factory($testName, '', '');

        // insert into the test table
        $DB->insert($obj->table, ['CommentID' => $commentID]);
```

Reply:
it doesnt #7155

Reply:
I think this thread is now resolved.